staging/src/k8s.io/api/resource/v1alpha3/types.go

```diff
@@ -0,0 +1,998 @@
+	//
+	// This is an alpha field and requires enabling the DRAControlPlaneController
+	// feature gate.
+	//
+	// +optional
+	ControllerName *string `json:"controllerName,omitempty" protobuf:"bytes,4,opt,name=controllerName"`
+
+	// Future extension, ignored by older schedulers. This is fine because
+	// scoring allows users to define a preference, without making it a
+	// hard requirement.
+	//
+	// Score *SomeScoringStruct
+}
+
+// Request is a request for one of many resources required for a claim.
+// This is typically a request for a single resource like a device, but can
+// also ask for several identical devices. It might get extended to support
+// asking for one of several different alternatives.
+type Request struct {
+	// The name can be used to reference this request in a pod.spec.containers[].resources.claims
+	// entry and in a constraint of the claim.
+	//
+	// Must be a DNS label.
+	Name string `json:"name" protobuf:"bytes,1,name=name"`
+
+	*RequestDetail `json:",inline" protobuf:"bytes,2,name=requestDetail"`
+
+	// FUTURE EXTENSION:
+	//
+	// OneOf contains a list of requests, only one of which must be satisfied.
+	// Requests are listed in order of priority.
+	//
+	// +optional
+	// +listType=atomic
+	// OneOf []RequestDetail
+}
+
+// RequestDetail is embedded inside Request. Exactly one field must be set.
+type RequestDetail struct {
+	// Device requests one or more devices.
+	//
+	// +required
+	Device *DeviceRequest `json:"device,omitempty" protobuf:"bytes,1,opt,name=device"`
+}
+
+// DeviceRequest is currently the only permitted alternative in RequestDetail.
+type DeviceRequest struct {
+	// By referencing a DeviceClass, a request inherits additional
+	// configuration parameters and selectors.
+	//
+	// A class is required. Which classes are available depends on the cluster.
+	//
+	// Administrators may use this to restrict which devices may get
+	// requested by only installing classes with selectors for permitted
+	// devices. If users are free to request anything without restrictions,
+	// then an empty class called "none" can get created to permit
+	// `deviceClassName: none`.
+	//
+	// +required
+	DeviceClassName string `json:"deviceClassName" protobuf:"bytes,1,name=deviceClassName"`
+
+	// Each selector must be satisfied by a device which is requested.
+	//
+	// +optional
+	// +listType=atomic
+	Selectors []Selector `json:"selectors,omitempty" protobuf:"bytes,2,name=selectors"`
+
+	// The count mode together with, for some modes, additional fields
+	// determines how many devices to allocate for the request.
+	//
+	// The default if unset is exactly one device:
+	//     countMode: Exact
+	//     count: 1
+	//
+	// "countMode: All" asks for all devices matching the selectors.
+	// Allocation fails if not all of them are available, unless admin
+	// access is requested. Admin access is granted also for
+	// devices which are in use.
+	//
+	// More modes may get added in the future.
+	//
+	// +default
```

Comment:
Ack.

Reply:
Hmm, `+default=ref(CountModeExact)` generates:
```
				if a.RequestDetail.Device.Count == nil {
					ptrVar1 := int64(v1alpha3.CountModeExact)
					a.RequestDetail.Device.Count = &ptrVar1
				}
```

That's broken because `CountModeExact` is a string.

I'm dropping the `+default` here in favor of hand-written code that implements the default if (and only if) the count mode is exact.

Reply:
Huh?  If you said `+default=ref(CountModeExact)` on the CountMode field, and it generated an int64, that's a huge bug.  If you put that tag on the Count field I could see it doing that...  Double check for me?

Reply:
I had mixed up the fields and had `+default="Exact"` for `CountMode` (works) and `+default=ref(CountModeExact)` for `Count`. So the `ref` works, I just didn't understand what it does until I tried it (incorrectly...).

I prefer `+default="Exact"` because that's clearer than referencing a constant.

Reply:
Resolving this thread, but the debate between a const and a literal in `+default` is not new.  I thought `ref()` was nice so we didn't have the same information in multiple places, but we do not yet have a clear convention.